PFCandidate no longer inherits from CompositeCandidate

Dr15Jones opened the pull request:
#### PR description:

The storing of PFCandidates takes 1/3rd of the entire time it takes to write AOD. I have started to look at the class to see if there was a way to make it easier for ROOT to store. The first think I tried to was to remove CompositeCandidate (as that looks hard to store) in order to see what would break and therefore would have to be changed if I removed the inheritance. To my surprise no code in CMSSW made use of the possible compositeness of the class.

#### PR validation:

The Code and all dependent code compiles.

Dr15Jones committed:
PFCandidate no longer inherits from CompositeCandidate

No code in CMSSW made use of the possible compositeness of the
class.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38999/31468

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- DataFormats/ParticleFlowCandidate (**reconstruction**)
- DataFormats/PatCandidates (**reconstruction**)


@jpata, @cmsbuild, @clacaputo can you please review it and eventually sign? Thanks.
@gouskos, @rovere, @lgray, @missirol, @hatakeyamak, @gpetruc this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

Dr15Jones commented:
I did a test where I wrote a 10,000 event file containing only `recoPFCandidates_particleFlow__RECO` using standard AOD settings. If I wrote the file with no compression, this PR made the output module run 40% faster. If I used the standard AOD compression, the times were the same within measurement error (just 1% difference).

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-INPUT AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/summary.html
**COMMIT**: 85e05422811d60165036aaa5d682a422cb506402
**CMSSW**: CMSSW_12_5_X_2022-08-08-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38999/26710/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test runtestTqafTopJetCombination had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
---> test runtestTqafTopTools had ERRORS
---> test runtestTqafTopHitFit had ERRORS
and more ...
</pre>

## RelVals

- [136.8311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrix-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017)```136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017/step2_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017.log```
- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM)```136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/step2_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM.log```
- [136.88811](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrix-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL)```136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL/step2_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [158.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrix-results/158.01_HydjetQ_reminiaodPbPb2018_INPUT+HydjetQ_reminiaodPbPb2018_INPUT+REMINIAODHI2018PPRECOMB+HARVESTHI2018PPRECOMINIAOD)
</details>



## RelVals-INPUT

- [136.72411](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.72411_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM)```136.72411_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM/step2_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM.log```
- [136.72412](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.72412_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMININANO_data2016UL_HIPM+HARVESTDR2_REMININANO_data2016UL_HIPM)```136.72412_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMININANO_data2016UL_HIPM+HARVESTDR2_REMININANO_data2016UL_HIPM/step2_RunJetHT2016B_reminiaodUL+RunJetHT2016B_reminiaodUL+REMININANO_data2016UL_HIPM+HARVESTDR2_REMININANO_data2016UL_HIPM.log```
- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM)```136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/step2_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [136.76111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.76111_RunJetHT2016E_reminiaodUL+RunJetHT2016E_reminiaodUL+REMINIAOD_data2016UL_HIPM+HARVESTDR2_REMINIAOD_data2016UL_HIPM)
- [136.7721](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.7721_RunJetHT2016H_reminiaod+RunJetHT2016H_reminiaod+REMINIAOD_data2016+HARVESTDR2_REMINIAOD_data2016)
- [136.77211](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.77211_RunJetHT2016H_reminiaodUL+RunJetHT2016H_reminiaodUL+REMINIAOD_data2016UL+HARVESTDR2_REMINIAOD_data2016UL)
- [136.83111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.83111_RunJetHT2017F_reminiaodUL+RunJetHT2017F_reminiaodUL+REMINIAOD_data2017UL+HARVEST2017_REMINIAOD_data2017UL)
- [136.8311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017)
- [136.8391](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.8391_RunExpressPhy2017F+RunExpressPhy2017F+HLTDR2_2017+RECODR2_2017reHLTSiPixelCalZeroBias_Prompt+HARVEST2017)
- [136.88811](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL)
- [140.5611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/140.5611_RunHI2018AOD+RunHI2018AOD+REMINIAODHID18+HARVESTHI18MINIAOD)
- [158.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/158.01_HydjetQ_reminiaodPbPb2018_INPUT+HydjetQ_reminiaodPbPb2018_INPUT+REMINIAODHI2018PPRECOMB+HARVESTHI2018PPRECOMINIAOD)
- [159.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/159.01_HydjetQ_reminiaodPbPb2021_INPUT+HydjetQ_reminiaodPbPb2021_INPUT+REMINIAODHI2021PPRECOMB+HARVESTHI2021PPRECOMINIAOD)
- [1325.51](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/1325.51_TTbar_13_94XreminiaodINPUT+TTbar_13_94XreminiaodINPUT+REMINIAOD_mc2017+HARVESTUP17_REMINIAOD_mc2017)
- [1325.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/1325.5_ProdZEE_13_reminiaodINPUT+ProdZEE_13_reminiaodINPUT+REMINIAOD_mc2016+HARVESTDR2_REMINIAOD_mc2016)
- [1325.516](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/1325.516_TTbar_13_reminiaod2016UL_preVFP_INPUT+TTbar_13_reminiaod2016UL_preVFP_INPUT+REMINIAOD_mc2016UL_preVFP+HARVESTDR2_REMINIAOD_mc2016UL_preVFP)
- [1325.5161](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/1325.5161_TTbar_13_reminiaod2016UL_postVFP_INPUT+TTbar_13_reminiaod2016UL_postVFP_INPUT+REMINIAOD_mc2016UL_postVFP+HARVESTDR2_REMINIAOD_mc2016UL_postVFP)
- [1325.517](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/1325.517_TTbar_13_reminiaod2017UL_INPUT+TTbar_13_reminiaod2017UL_INPUT+REMINIAOD_mc2017UL+HARVESTUP17_REMINIAOD_mc2017UL)
- [1325.518](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/runTheMatrixINPUT-results/1325.518_TTbar_13_reminiaod2018UL_INPUT+TTbar_13_reminiaod2018UL_INPUT+REMINIAOD_mc2018UL+HARVESTUP18_REMINIAOD_mc2018UL)
</details>



## AddOn Tests

- [pat1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/26710/addOnTests/pat1)```cmsRun /data/cmsbld/jenkins/workspace/ib-run-pr-addon/CMSSW_12_5_X_2022-08-08-1100/src/PhysicsTools/PatAlgos/test/IntegrationTest_cfg.py : FAILED - time: date Mon Aug  8 21:01:15 2022-date Mon Aug  8 20:59:52 2022 s - exit: 35584```

Dr15Jones commented:
@pcanal So I changed the inheritance for `reco::PFCandidate`, dropping one of the intermediate base classes. When I did that I updated all the class versions as well. However, now the tests run for the PR are failing where all these tests seem to be reading back `reco::PFCandidate`s which had been stored using the previous version. It seems like the scheme evolution isn't working properly.

Dr15Jones commented:
So I wrote a simple EDAnalyzer which dumps the contents of a `std::vector<reco::PFCandidate>`. I ran it using both the original code and the new code, both reading the same input file (which was written using the original format). The results are

```
element 0	PFCandidate type: 1 E/pT/eta/phi 0.811/0.175/-2.196/-0.403, blocks/ie |	element 0	PFCandidate type: 1 E/pT/eta/phi 0.000/0.175/-2.196/-0.403, blocks/ie
element 1	PFCandidate type: 1 E/pT/eta/phi 1.500/0.230/2.559/2.968, blocks/iele |	element 1	PFCandidate type: 1 E/pT/eta/phi 0.000/0.230/2.559/2.968, blocks/iele
element 2	PFCandidate type: 1 E/pT/eta/phi 0.589/0.180/1.825/-2.992, blocks/iel |	element 2	PFCandidate type: 1 E/pT/eta/phi 0.000/0.180/1.825/-2.992, blocks/iel
element 3	PFCandidate type: 1 E/pT/eta/phi 1.244/0.205/2.484/-3.059, blocks/iel |	element 3	PFCandidate type: 1 E/pT/eta/phi 0.000/0.205/2.484/-3.059, blocks/iel
element 4	PFCandidate type: 1 E/pT/eta/phi 9.328/2.029/-2.206/-1.708, blocks/ie |	element 4	PFCandidate type: 1 E/pT/eta/phi 0.000/2.029/-2.206/-1.708, blocks/ie
element 5	PFCandidate type: 1 E/pT/eta/phi 2.963/0.630/-2.229/-0.503, blocks/ie |	element 5	PFCandidate type: 1 E/pT/eta/phi 0.000/0.630/-2.229/-0.503, blocks/ie
```

So when reading back using the new format, energy is always 0 while pT, eta, and phi look fine.

pcanal commented:
Which data member does the energy and pT (or eta or phi) correspond to?

Dr15Jones commented:
So `PFCandidate::energy()` is read from `LeafCandidate::m_state:: p4Cartesian_` where that member variable is declared _transient_ to ROOT. 

https://github.com/cms-sw/cmssw/blob/e52f99a59bb77ceec0102769446b26c321662ae8/DataFormats/Candidate/src/classes_def.xml#L7

the value is supposed to be filled by the IO rule
https://github.com/cms-sw/cmssw/blob/e52f99a59bb77ceec0102769446b26c321662ae8/DataFormats/Candidate/src/classes_def.xml#L9-L12

So it appears that when schema evolution is being used, that the necessary IO rules are not always being run.

Dr15Jones commented:
So pT, eta and phi come from `LeafCandidate::m_state::p4Polar_` which is stored.

Dr15Jones commented:
So if I do
```
[cdj@cmslpc-c8-heavy01 matrix]$ root -l root://eoscms.cern.ch//eos/cms/store/user/cmsbuild/store/data/Run2017F/JetHT/AOD/17Nov2017-v1/70000/0054D144-36DF-E711-B074-02163E01A47B.root
root [1] Events->Scan("recoPFCandidates_particleFlow__RECO.obj.m_state.energy()")
```

using the original format I get
```
***********************************
*    Row   * Instance * recoPFCan *
***********************************
*        0 *        0 * 0.8107478 *
*        0 *        1 * 1.5003893 *
*        0 *        2 * 0.5893012 *
*        0 *        3 * 1.2443912 *
*        0 *        4 * 9.3277593 *
*        0 *        5 * 2.9629352 *
*        0 *        6 * 3.7514724 *
*        0 *        7 * 2.8305396 *
*        0 *        8 * 1.6818398 *
*        0 *        9 * 1.7144627 *
*        0 *       10 * 0.9533503 *
*        0 *       11 * 7.2343502 *
*        0 *       12 * 5.6262460 *
*        0 *       13 * 0.9747063 *
*        0 *       14 * 9.3023527 *
*        0 *       15 * 1.4282982 *
*        0 *       16 * 4.3854562 *
*        0 *       17 * 1.2034772 *
*        0 *       18 * 1.9004409 *
*        0 *       19 * 1.6507877 *
*        0 *       20 * 1.9658588 *
*        0 *       21 * 1.9287844 *
*        0 *       22 * 2.9119759 *
*        0 *       23 * 0.7939654 *
*        0 *       24 * 0.6640725 *
```

but using the new format I get
```
***********************************
*    Row   * Instance * recoPFCan *
***********************************
*        0 *        0 *         0 *
*        0 *        1 *         0 *
*        0 *        2 * 2.84e-306 *
*        0 *        3 *         0 *
*        0 *        4 *         0 *
*        0 *        5 *         0 *
*        0 *        6 *         0 *
*        0 *        7 *         0 *
*        0 *        8 * 2.84e-306 *
*        0 *        9 * 2.84e-306 *
*        0 *       10 * 2.84e-306 *
*        0 *       11 *         0 *
*        0 *       12 *         0 *
*        0 *       13 * 2.84e-306 *
*        0 *       14 *         0 *
*        0 *       15 * 2.84e-306 *
*        0 *       16 *         0 *
*        0 *       17 * 2.84e-306 *
*        0 *       18 *         0 *
*        0 *       19 * 2.84e-306 *
*        0 *       20 *         0 *
*        0 *       21 * 2.84e-306 *
*        0 *       22 *         0 *
*        0 *       23 * 3.72e-307 *
*        0 *       24 * 2.84e-306 *
```

pcanal commented:
Fair enough.  It seems to have to do with the handling/passing-through to the base class.  Let me try to write a standalone reproducer.

pcanal commented:
I can reproduce locally.  I am looking for a solution.

mandrenguyen commented:
@pcanal @Dr15Jones Just wondering if there's any news here

pcanal commented:
I am mking steady progress towards the fix.

mandrenguyen commented:
@pcanal @Dr15Jones 
Any news on this?

cmsbuild commented:
Pull request #38999 was updated. @simonepigazzini, @mandrenguyen, @clacaputo, @vlimant can you please check and sign again.

mandrenguyen commented:
-reconstruction
Cleaning the reco queue

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-INPUT AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/summary.html
**COMMIT**: 85e05422811d60165036aaa5d682a422cb506402
**CMSSW**: CMSSW_13_3_X_2023-10-22-0000/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38999/35338/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 8 errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
and more ...
</pre>

## RelVals

- [136.88811](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrix-results/136.88811_RunJetHT2018DreMINIAODUL)```136.88811_RunJetHT2018DreMINIAODUL/step2_RunJetHT2018DreMINIAODUL.log```
- [136.8311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrix-results/136.8311_RunJetHT2017FreMINIAOD)```136.8311_RunJetHT2017FreMINIAOD/step2_RunJetHT2017FreMINIAOD.log```
- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrix-results/136.7611_RunJetHT2016EreMINIAOD)```136.7611_RunJetHT2016EreMINIAOD/step2_RunJetHT2016EreMINIAOD.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [158.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrix-results/158.01_HydjetQB12ppRECOin2018reMINIAOD)
- [2500.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrix-results/2500.4_NANOmc132X)
</details>



## RelVals-INPUT

- [136.72411](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.72411_RunJetHT2016B_reminiaodUL)```136.72411_RunJetHT2016B_reminiaodUL/step2_RunJetHT2016B_reminiaodUL.log```
- [136.72412](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.72412_RunJetHT2016B_reminiaodUL)```136.72412_RunJetHT2016B_reminiaodUL/step2_RunJetHT2016B_reminiaodUL.log```
- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.7611_RunJetHT2016EreMINIAOD)```136.7611_RunJetHT2016EreMINIAOD/step2_RunJetHT2016EreMINIAOD.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [136.76111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.76111_RunJetHT2016EreMINIAODUL)
- [136.8391](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.8391_RunExpressPhy2017F)
- [136.7721](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.7721_RunJetHT2016HreMINIAOD)
- [136.77211](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.77211_RunJetHT2016HreMINIAODUL)
- [136.8311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.8311_RunJetHT2017FreMINIAOD)
- [136.83111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.83111_RunJetHT2017FreMINIAODUL)
- [136.88811](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/136.88811_RunJetHT2018DreMINIAODUL)
- [140.201](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/140.201_RunJetMET2022D_reMINI)
- [140.5611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/140.5611_RunHI2018reMINIAOD)
- [140.202](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/140.202_RunJetMET2022D_reMINI)
- [159.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/159.01_HydjetQ_reminiaodPbPb2022_INPUT)
- [158.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/158.01_HydjetQB12ppRECOin2018reMINIAOD)
- [1325.51](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/1325.51_TTbar13reMINIAOD)
- [1325.517](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/1325.517_TTbar13reMINIAODin2017UL)
- [1325.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/1325.5_ZEE13reMINIAOD)
- [1325.5161](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/1325.5161_TTbar13reMINIAODin2016ULpostVFP)
- [1325.516](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/1325.516_TTbar13reMINIAODin2016ULpreVFP)
- [1325.518](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/1325.518_TTbar13reMINIAODin2018UL)
- [2500.2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.2_NANOmc124Xrun3)
- [2500.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.1_NANOmc122Xrun3)
- [2500.3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.3_NANOmc130X)
- [2500.31](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.31_NANOdata130Xrun3)
- [2500.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.21_NANOdata124Xrun3)
- [2500.311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.311_NANOdata130Xrun3)
- [2500.211](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.211_NANOdata124Xrun3)
- [2500.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/runTheMatrixINPUT-results/2500.4_NANOmc132X)
</details>



## AddOn Tests

- [pat1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/35338/addOnTests/pat1)
```
[pat1:1] cmsRun /data/cmsbld/jenkins/workspace/ib-run-pr-addon/CMSSW_13_3_X_2023-10-22-0000/src/PhysicsTools/PatAlgos/test/IntegrationTest_cfg.py : FAILED - elapsed time: 48 sec (ended on Sun Oct 22 23:42:46 2023) - exit: 35584
```

smuzaffar commented:
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

Dr15Jones commented:
@pcanal was this problem solved?

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_1_X. Please open a backport if it should also go in to CMSSW_14_0_X.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_2_X. Please open a backport if it should also go in to CMSSW_14_1_X.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_1_X. Please open a backport if it should also go in to CMSSW_15_0_X.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals RelVals-INPUT AddOn 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/summary.html
**COMMIT**: 85e05422811d60165036aaa5d682a422cb506402
**CMSSW**: CMSSW_15_0_X_2025-02-06-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38999/44258/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 8 errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test testRecoMETMETProducers had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
and more ...
</pre>

## RelVals

- [2500.201](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrix-results/2500.201_NANOmc140X)
```
----- Begin Fatal Exception 07-Feb-2025 10:34:43 CET-----------------------
An exception of category 'FileReadError' occurred while
   [0] Processing  Event run: 1 lumi: 85 event: 8402 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module NanoAODDQM/'nanoDQMMC'
   [3] Prefetching for module TriggerObjectTableProducer/'triggerObjectTable'
   [4] While reading from source BXVector<l1t::EGamma> caloStage2Digis 'EGamma' RECO
   [5] Rethrowing an exception that happened on a different read request.
   [6] Processing  Event run: 1 lumi: 85 event: 8402 stream: 0
   [7] Running path 'dqmoffline_step'
   [8] Prefetching for module NanoAODDQM/'nanoDQMMC'
   [9] Prefetching for module CandMCMatchTableProducer/'boostedTauMCTable'
   [10] Prefetching for module PATObjectCrossLinker/'linkedObjects'
   [11] Prefetching for module PATMuonRefSelector/'finalMuons'
   [12] Prefetching for module PATMuonUserDataEmbedder/'slimmedMuonsWithUserData'
   [13] Prefetching for module PATMuonUpdater/'slimmedMuonsUpdated'
   [14] While reading from source std::vector<pat::Muon> slimmedMuons '' RECO
   [15] Reading branch patMuons_slimmedMuons__RECO.
   Additional Info:
      [a] Fatal Root Error: @SUB=TBufferFile::ReadClassBuffer
Could not find the StreamerInfo for version 11 of the class edm::RefVector<vector<reco::PFBlock>,reco::PFBlock,edm::refhelper::FindUsingAdvance<vector<reco::PFBlock>,reco::PFBlock> >, object skipped at offset 27231

----- End Fatal Exception -------------------------------------------------
```


## RelVals-INPUT

- [2500.232](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.232_EGMNANOdata140Xrun3)```2500.232_EGMNANOdata140Xrun3/step2_EGMNANOdata140Xrun3.log```
- [2500.233](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.233_BTVNANOdata140Xrun3)```2500.233_BTVNANOdata140Xrun3/step2_BTVNANOdata140Xrun3.log```
- [2500.231](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.231_muPOGNANO140Xrun3)```2500.231_muPOGNANO140Xrun3/step2_muPOGNANO140Xrun3.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [2500.236](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.236_lepTrackInfoNANOdata140Xrun3)
- [2500.234](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.234_jmeNANOdata140Xrun3)
- [2500.224](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.224_jmeNANOmc140X)
- [2500.311](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.311_NANOdata2024reMINI)
- [2500.112](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.112_NANOdata130Xrun3)
- [2500.238](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.238_ScoutingNANOwithPromptdata140Xrun3)
- [2500.211](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.211_NANOdata140Xrun3)
- [2500.212](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.212_NANOdata140Xrun3)
- [159.01](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/159.01_HydjetQ_reminiaodPbPb2022_INPUT)
- [2500.031](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.031_NANOdataUL16APVreMINI)
- [2500.111](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.111_NANOdata130Xrun3)
- [2500.033](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.033_NANOdataUL17reMINI)
- [2500.235](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.235_jmeNANOrePuppidata140Xrun3)
- [136.8391](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/136.8391_RunExpressPhy2017F)
- [140.5611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/140.5611_RunHI2018reMINIAOD)
- [2500.225](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.225_jmeNANOrePuppimc140X)
- [2500.034](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.034_NANOdataUL18reMINI)
- [2500.226](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.226_lepTrackInfoNANOmc140X)
- [2500.221](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.221_muPOGNANOmc140X)
- [2500.223](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.223_BTVNANOmc140X)
- [2500.222](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.222_EGMNANOmc140X)
- [2500.032](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.032_NANOdataUL16reMINI)
- [2500.228](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.228_ScoutingNANOwithPromptmc140X)
- [2500.201](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.201_NANOmc140X)
- [2500.301](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.301_NANOmc2024reMINI)
- [2500.022](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.022_NANOmcUL16reMINI)
- [2500.024](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.024_NANOmcUL18reMINI)
- [2500.023](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.023_NANOmcUL17reMINI)
- [2500.101](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.101_NANOmc130X)
- [2500.021](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/runTheMatrixINPUT-results/2500.021_NANOmcUL16APVreMINI)
</details>



## AddOn Tests

- [pat1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5d55e/44258/addOnTests/pat1)
```
A fatal system signal has occurred: segmentation violation
```

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_16_0_X. Please open a backport if it should also go in to CMSSW_15_1_X.

makortel commented:
This PR was moved to https://github.com/cms-sw/cmssw/pull/50682 as part of the EVOLUTION_X work
